planetiler-core/src/main/java/com/onthegomap/planetiler/geo/TileExtents.java

```diff
@@ -52,14 +85,35 @@ public boolean test(TileCoord tileCoord) {
 
   /**
    * X/Y extents within a given zoom level. {@code minX} and {@code minY} are inclusive and {@code maxX} and {@code
-   * maxY} are exclusive.
+   * maxY} are exclusive. shape is an optional polygon defining a more refine shape
    */
-  public record ForZoom(int minX, int minY, int maxX, int maxY) {
+  public record ForZoom(int minX, int minY, int maxX, int maxY, PreparedGeometry shape) {
 
     public boolean test(int x, int y) {
+      if (shape != null) {
+
+        return shape.intersects(JTS_FACTORY.toGeometry(new Envelope(
+          x,
+          x + 1.0f,
+          y,
+          y + 1.0f
+        )));
+      }
       return testX(x) && testY(y);
     }
 
+    public boolean testOverShape(int x, int y, int z) {
```

Comment:
@msbarry i did that to prevent another test on `shape != null`. in `TiledGeometry` we only want to test if shape exists.
But in `test` we only want to test if `shape` too. So it would mean testing for shape twice

Reply:
actually i was doing the test for shape in `testOverShape` :s Is that ok if i remove the test for `shape!=null` in `testOverShape` ? 
Or do you prefer to use `test` but then have 2 tests `shape!=null`

Reply:
@msbarry if i use `test` then test will always be called even if no shape.

Reply:
Yes, I think that's fine though since if shape is null it will just do a range check on x/y which should be cheap/free